std/algorithm/comparison.d

```diff
@@ -780,7 +795,8 @@ template equal(alias pred = "a == b")
                 if (r2.empty) return false;
                 if (!binaryFun!(pred)(r1.front, r2.front)) return false;
             }
-            return r2.empty;
+            static if (!isInfinite!Range1)
+                return r2.empty;
```

Comment:
If Range1 is infinite, r1.empty is false, so the for loop condition is true, so without the static if the return line causes an unreachable code warning. Lines 904 and 905 test what you're asking for.